text/3660-crates-io-crate-deletions.md

```diff
@@ -0,0 +1,123 @@
+- Start Date: 2024-06-20
+- RFC PR: [rust-lang/rfcs#3660](https://github.com/rust-lang/rfcs/pull/3660)
+
+# Summary
+
+This RFC proposes a mechanism for crate authors to delete their crates from crates.io under certain conditions.
+
+
+# Motivation
+
+There are a variety of reasons why a crate author might want to delete a crate or version from crates.io:
+
+* You published something accidentally.
+* You wanted to test crates.io.
+* You published content you didn't intend to be public.
+* You want to rename a crate. (The only way to rename a package is to re-publish it under a new name)
+
+The current [crates.io usage policy](https://crates.io/policies) says:
+
+> Crate deletion by their owners is not possible to keep the registry as immutable as possible.
+
+This restriction makes sense for the majority of crates that have been around for a while and are actively used, but the above list of reasons shows that there are valid use cases for allowing crate authors to delete their crates without having to contact the crates.io team.
+
+To make this process easier for our users and to reduce the workload of the crates.io team dealing with such support requests, we propose to codify our current set of informal rules into a formal policy that allows crate authors to delete their crates themselves under certain conditions (see below).
+
+
+# Proposal
+
+We propose to allow crate authors to delete their **crates** from crates.io under the following conditions:
+
+* The crate has been published for less than 72 hours,
```

Comment:
Thoughts on allowing the publisher to define the timeframe? I think the goal of 72 hours is to give the user time to discover their mistake, but not all mistakes are found so quickly.
We would obviously want the user-supplied length of time to be visible to users of the crate, so they are aware of the potential for removal.
This gives the crate owner a window that they are comfortable with while also giving crate users a clear understanding of the risk that the crate may be removed from under them.

Reply:
> Thoughts on allowing the publisher to define the timeframe?

It is technically possible, but I'm not sure it's worth the complexity. Rust users would have to keep track of all the different timeframes in their transitive dependencies to be able to reason about deletions.

Reply:
I now see how that would add quite a lot of complexity for users, forcing them to understand if any dependency in the hierarchy (parent-to-child) back to the leaf dependencies have the potential to be deleted from under them as they try to understand the risk of using a single crate that may depend on many crates (and then those may depend on many crates, etc.).
Sticking with a static, short time frame makes a lot more sense given that.